### PR TITLE
Add endpoint for downloading a project

### LINKED ISSUE #7194: Project Manager endpoint for downloading a project as an `.enso-project` archive
Required for [cloud-v2/#478](https://github.com/enso-org/cloud-v2/issues/478).

It should accept a `projectId`, and return either:
- the archive, encoded as a JSON string literal, or
- the full URL to a temporary HTTP port (WebSocket is fine too, although unnecessary because the client shouldn't be sending), or
- any other method

The important part is that it shouldn't write to the filesystem, since `gui watch`, and `localhost:8080` when using the IDE's builtin server, both do not have access to the FS.

The ideal solution would allow piping the response body to cloud, to allow easy uploading of projects to the cloud backend.
(A HTTP response should almost certainly allow this, by simply passing `response.body` to the `fetch` request to the backend. Other methods like WebSocket would probably work too, just much less trivial)

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #7194

Changelog:

- add: `/projects/{project_id}/enso_project` HTTP endpoint returning an `.enso-project` archive structure
- update: archive enso project to a `.enso-project` `.tar.gz` archive
- update: make project `path` a required field

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchive.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorOutputStream;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/** An enso project in .enso-project format. */
+public record EnsoProjectArchive(ByteArrayOutputStream outputStream) {
```
Comment (JaroslavTulach) [inline lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchive.java]: `record` is by definition immutable, but here it has a _heavily mutable_ field. I would rather expose:
- `byte[] getContent()` providing (a copy?) of the stored content
- `InputStream getContent()` providing a way to read the content via stream

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/CloseArchiveEntryException.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package org.enso.pkg.archive;
+
+import java.io.IOException;
+
+/** Error when trying to close an enso-project archive entry. */
+public final class CloseArchiveEntryException extends EnsoProjectArchiveException {
```
Comment (JaroslavTulach) [inline lib/scala/pkg/src/main/java/org/enso/pkg/archive/CloseArchiveEntryException.java:6]: This class is `public`, but it extends non-`public` class. That's not a nice design.

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchiveException.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package org.enso.pkg.archive;
+
+import java.io.IOException;
+
+/** Base class for errors when building enso-project archive. */
+abstract class EnsoProjectArchiveException extends IOException {
+
+  public EnsoProjectArchiveException(IOException cause) {
```
Comment (JaroslavTulach) [inline lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchiveException.java:8]: Why do we need to wrap an `IOException` with a subclass of `IOException`? Can't we just propagate `IOException` directly? Just like `EnsoProjectArchive build(...) throws IOException` does?

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/TarArchiveBuildingFileVisitor.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import java.io.InputStream;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.apache.commons.compress.archivers.tar.TarConstants;
+import org.enso.pkg.Package$;
+
+/** A file visitor that builds a tar archive out of visited files. */
+public final class TarArchiveBuildingFileVisitor implements FileVisitor<Path> {
```
Comment (JaroslavTulach) [inline lib/scala/pkg/src/main/java/org/enso/pkg/archive/TarArchiveBuildingFileVisitor.java]: I'd expect this to be an implementation detail class and I'd keep it _package private_.

File: lib/scala/pkg/src/test/scala/org/enso/pkg/archive/EnsoProjectArchiveSpec.scala [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  TarArchiveInputStream
+}
+import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+import java.io.ByteArrayInputStream
+import java.nio.file.Path
+
+class EnsoProjectArchiveSpec extends AnyWordSpec with Matchers {
```
Comment (JaroslavTulach) [inline lib/scala/pkg/src/test/scala/org/enso/pkg/archive/EnsoProjectArchiveSpec.scala:13]: Eeeh! New file in Scala.

File: lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/http/ProjectsEndpoint.scala [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+          "Failed to build an enso-project archive [{}]. {}",
+          projectId,
+          err
+        )
+        complete(StatusCodes.InternalServerError)
+      case Success(Right(None)) =>
+        complete(StatusCodes.NotFound)
+      case Success(Right(Some(archive))) =>
+        complete(
+          HttpEntity(
```
Comment (JaroslavTulach) [inline lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/http/ProjectsEndpoint.scala:65]: Shouldn't we also write a test for this endpoint?

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchiveException.java (hunk 3 above)
Comment (4e6) [reply lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchiveException.java:8]: I wanted to distinguish those from other IO exceptions

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchive.java (hunk 1 above)
Comment (4e6) [reply lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchive.java]: Changed to contain the resulting byte array

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/CloseArchiveEntryException.java (hunk 2 above)
Comment (4e6) [reply lib/scala/pkg/src/main/java/org/enso/pkg/archive/CloseArchiveEntryException.java:6]: Updated

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/TarArchiveBuildingFileVisitor.java (hunk 4 above)
Comment (4e6) [reply lib/scala/pkg/src/main/java/org/enso/pkg/archive/TarArchiveBuildingFileVisitor.java]: Made package private

File: lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/http/ProjectsEndpoint.scala (hunk 6 above)
Comment (4e6) [reply lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/http/ProjectsEndpoint.scala:65]: All the tricky logic is about the archive building. This one is pretty straightforward. I'll see if I can mock the project repository and add a test then

File: lib/scala/pkg/src/test/scala/org/enso/pkg/archive/EnsoProjectArchiveSpec.scala (hunk 5 above)
Comment (4e6) [reply lib/scala/pkg/src/test/scala/org/enso/pkg/archive/EnsoProjectArchiveSpec.scala:13]: Sneaked one in :hand_over_mouth:

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchiveException.java (hunk 3 above)
Comment (JaroslavTulach) [reply lib/scala/pkg/src/main/java/org/enso/pkg/archive/EnsoProjectArchiveException.java:8]: Distinguishing is OK. I also  created something like `EnsoProjectArchiveException(String msg)` exception that extends `IOException`. I just don't understand why it takes `IOException` as a delegate... but now I think that's OK.

File: lib/scala/pkg/src/main/java/org/enso/pkg/archive/CloseArchiveEntryException.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+  public CloseArchiveEntryException(IOException cause) {
```
Comment (JaroslavTulach) [inline lib/scala/pkg/src/main/java/org/enso/pkg/archive/CloseArchiveEntryException.java:8]: If the class (and other similar cases) is only constructed in the same package, then you may want to have _package private_ constructors.

**Review by JaroslavTulach — APPROVED** [review]

**Pull request merged**